src/tests/JIT/Regression/JitBlue/Runtime_54118/Runtime_54118.cs

```diff
@@ -0,0 +1,239 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+
+// Various tests for memory-dependent loop hoisting
+
+class Runtime_54118
+{
+    public static int Main()
+    {
+        _clsVar = -1;
+        int result = 0;
+        int index = 0;
+        void Test(string name, Func<bool> act)
+        {
+            Console.Write("{0}: ", name);
+            if (act())
+            {
+                Console.WriteLine("PASS");
+            }
+            else
+            {
+                Console.WriteLine("FAIL");
+                result |= 1 << index;
+            }
+
+            index++;
+        }
+
+        Test(nameof(TestConstantByref), TestConstantByref);
+        Test(nameof(TestConstantArr), TestConstantArr);
+        Test(nameof(TestConstantClsVar), TestConstantClsVar);
+        Test(nameof(TestParamByref), () => TestParamByref(1));
+        Test(nameof(TestParamArr), () => TestParamArr(1));
+        Test(nameof(TestParamClsVar), () => TestParamClsVar(1));
+        Test(nameof(TestPhiByref), TestPhiByref);
+        Test(nameof(TestPhiArr), TestPhiArr);
+        Test(nameof(TestPhiClsVar), TestPhiClsVar);
+        Test(nameof(TestCastByref), TestCastByref);
+        Test(nameof(TestCastArr), TestCastArr);
+        Test(nameof(TestCastClsVar), TestCastClsVar);
+
+        return 100 + result;
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static bool TestConstantByref()
+    {
+        int[] arr = { -1 };
+        ref int r = ref arr[0];
+        int val = -1;
+        for (int i = 0; i < 2; i++)
```

Comment:
To test loop-dependent VN, these tests all depend on these loops remaining as loops. But Checked JIT loop unrolling under stress will probably fully unroll them. And what if we change the loop unrolling heuristics? Maybe the upper bound should be make an argument.

Reply:
See #56184